### ISSUE TITLE
Improve page loading by caching map display with a static image

### ISSUE THREAD

**Issue by lordyavin** [theCrag/website#3108]
Recently I recognized that e.g. the map of world is loading quite slow. Of course this is caused by the amount of information we put into and transfer to google and back to the client. Today I remembered a simple solution we implemeneted in a quite different scenario but with equal problems.

_Why not using static images as preview of the default view port until the user wants to zoom or drag the map?
The preview image could be created and updated each time an editor updates the geolocation of the current node or one of its direct descendants (could imply a bottom-up-recursion that eats up all CPU power and increases disk IO)  or by a cron job. If no image is available fall back to google maps._

This is just an idea that worked for me in the past to avoid rendering stuff that isn't in the current focus of the user. I have no idea if this can be implemented for a website. One thing is 

@brendanheywood encouraged me in issue #3055 to open this one.

**Comment by brendanheywood** [comment]
Just to be clear, we don't transfer anything to google and back to the client, and the amount of geo data is surprisingly modest at 93k. The bulk of the data is map tiles which won't substantially change by making it one large static image. I really want to understand exactly what you are experiencing and why before talking about any solutions.

Can you please give me the exact url of the page (ie world? https://www.thecrag.com/maps) and what device / network are you on? A video would be fantastic. You also haven't defined 'slow', like 10 seconds? 30 seconds?

For the record (give we are about to rewrite this from scratch), right now on simulated 3G I'm getting ~4 second loads on a cold cache which while isn't brilliant is not what I would call slow either. Most of this is latency bound rather than io or cpu and finishing the move to http2 (https://github.com/theCrag/website/issues/2106) would probably help most here.

On my iphone with wifi turned off I'm loading in ~3 seconds.

But what you are describing feels to me more like it's cpu which is why I'd like to see a video, or maybe try and replicate it using a particular device / browser / network on https://www.webpagetest.org

If it is cpu bound, then probably the best solution is reducing the number of markers on the map before they get clustered, and the new approach will do exactly that so the client side render stage should be much faster. For reference the world map currently loads roughly 6500 markers which gets reduced into ~300 clusters, the new setup should get those two numbers much closer together, like 1000 -> 300

**Comment by lordyavin** [comment]
Okay here some more details. First a screenshot of the results of an Firefox Addon (Page Speed Monitor) that measures the loading time:

![grafik](https://user-images.githubusercontent.com/1528540/37149153-461ba7ea-22cd-11e8-9732-dd27ccaa37d9.png)

Screenshot of the result summary of webpagetest.org from: Frankfurt, Germany - EC2 - Firefox - Cable
8.3.2018, 12:38:44
![grafik](https://user-images.githubusercontent.com/1528540/37149507-8bc09f70-22ce-11e8-892b-7c0c114a1c45.png)


Unfortunately I don't have a tool to capture a video of the loading process. I'm behind a proxy and maybe this is the bottleneck. But I think the page should load as fast as possible,

**Comment by brendanheywood** [comment]
Right, ok I think the #1 issue here is your proxy, mostly probably via worse latency.

But that webpage test result still doesn't make sense yet, I just did the same test as you and a few times and I'm consistently getting load times around 4s, nothing like the 14 seconds you are getting.

https://www.webpagetest.org/result/180308_TX_502ab97e4c0824bdacf64d95d53a7abe/
https://www.webpagetest.org/result/180308_ND_176f5e276815c296a512142a42edad76
https://www.webpagetest.org/result/180308_58_09a4c600cbe8db99a4dcea9a1f5724c3/

**Comment by lordyavin** [comment]
17.889s https://www.webpagetest.org/result/180308_1E_901edf0765e78ea9a742417a60c3c1bc/14.440s https://www.webpagetest.org/result/180308_5B_c6fb1c74c7712b0db1514ef10c5142d9/

  Brendan Heywood <notifications@github.com> hat am 8. März 2018 um 14:44 geschrieben:
 
 
 Right, ok I think the #1 issue here is your proxy, mostly probably via worse latency. 
 But that webpage test result still doesn't make sense yet, I just did the same test as you and a few times and I'm consistently getting load times around 4s, nothing like the 14 seconds you are getting. 
 https://www.webpagetest.org/result/180308_TX_502ab97e4c0824bdacf64d95d53a7abe/ https://www.webpagetest.org/result/180308_ND_176f5e276815c296a512142a42edad76 https://www.webpagetest.org/result/180308_58_09a4c600cbe8db99a4dcea9a1f5724c3/

**Comment by brendanheywood** [comment]
aarggh, we are not even looking at the same page, this is why I asked for the exact url  :confused: 

That page doesn't even have google maps on it at all!

Yes that page was slow and it was the facebook embed on the right casuing the slow down, which previously was also on the dashboard and we have removed it from there so I have also removed it from here too. This it hot fixed.

I have re-run those exact tests of yours above and the load time is significantly better now:

Yours:

Firefox:
14.202s https://www.webpagetest.org/result/180308_5B_c6fb1c74c7712b0db1514ef10c5142d9/

Chrome
10.482s https://www.webpagetest.org/result/180308_7Z_9ad4633302a45deb97d99ece8da82a46/

Now:

Firefox: 
5.041s https://www.webpagetest.org/result/180309_1T_9638fae610b8d17568a9077e5f315145/

Chrome:
2.963s https://www.webpagetest.org/result/180309_27_9d45a8c0e170290e4a5ef5781932522d/

I still don't get why firefox is significantly slower, but I don't want to read too much into that
